### PR TITLE
[common] Support modification of bigint type to string type.

### LINKED ISSUE #1273: [Feature] Support modification of bigint type to string type
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

I need to change the field type of a table from bigint to string, which is currently not supported, and the following exception is reported: 
java.lang.IllegalStateException: Column type role_uid[BIGINT] cannot be converted to STRING without loosing information.

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### LINKED ISSUE #1276: [Feature] [core] Supports all explicit casting when changing column types
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Supports all explicit casting when changing column types, For example: string to int, long ...

related issue: #1273
related pr: #1274 

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by zhuangchong**
### Purpose

close #1273
close #1276 

### Tests

SchemaEvolutionTest#testUpdateFieldType
CastExecutorTest#testNumericToString

### CODE REVIEW

**Comment by JingsongLi** [comment]
Thanks @zhuangchong , looks good!~ Can you create an issue to supports all explicit casting?

**Comment by zhuangchong** [comment]
> Thanks @zhuangchong , looks good!~ Can you create an issue to supports all explicit casting?

Already supports all explicit casting.

https://github.com/apache/incubator-paimon/blob/998820f6576ddfa0915000d713a65cae47f0ff18/paimon-core/src/main/java/org/apache/paimon/schema/SchemaManager.java#L285-L292

**Comment by JingsongLi** [comment]
> Already supports all explicit casting.

String to int, long?

**Comment by zhuangchong** [comment]
> > Already supports all explicit casting.
> 
> String to int, long?

Sorry, I overlooked the string to number type conversion, currently supports number to string and number to number (bigint to int ...)

File: paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java [hunk 1]
```diff
@@ -97,6 +97,8 @@ public class CastExecutors {
                                         }
                                 }
                             };
+                        case VARCHAR:
+                            return Object::toString;
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java:101]: Return type should be `BinaryString`, not `String`. See the following case branches about `CHAR` and `VARCHAR`.

File: paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java [hunk 2]
```diff
@@ -123,6 +125,8 @@ public class CastExecutors {
                                             (Decimal) value,
                                             decimalType.getPrecision(),
                                             decimalType.getScale());
+                        case VARCHAR:
+                            return Object::toString;
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java:129]: Same as above.

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaManager.java [hunk 3]
```diff
@@ -283,7 +283,7 @@ public TableSchema commitChanges(List<SchemaChange> changes) throws Exception {
                             update.fieldName(),
                             (field) -> {
                                 checkState(
-                                        DataTypeCasts.supportsImplicitCast(
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/schema/SchemaManager.java:286]: By allowing explicit type casting you will allow casting from `string` to `bigint`, or `timestamp` to `tinyint`, etc., etc. Have you implemented all these in `CastExecutors`?

File: paimon-core/src/test/java/org/apache/paimon/casting/CastExecutorTest.java [hunk 4]
```diff
@@ -144,6 +144,42 @@ public void testDecimalToNumeric() {
                 1.23D);
     }
 
+    @Test
+    public void testNumericToString() {
```
Comment (tsreaper) [inline paimon-core/src/test/java/org/apache/paimon/casting/CastExecutorTest.java:148]: Cast result should be `BinaryString`, not normal `String`. See `testStringToString`.

File: paimon-core/src/test/java/org/apache/paimon/catalog/CatalogTestBase.java [hunk 5]
```diff
@@ -603,11 +603,11 @@ public void testAlterTableUpdateColumnType() throws Exception {
                                         identifier,
                                         Lists.newArrayList(
                                                 SchemaChange.updateColumnType(
-                                                        "col1", DataTypes.STRING())),
+                                                        "col1", DataTypes.DATE())),
```
Comment (tsreaper) [inline paimon-core/src/test/java/org/apache/paimon/catalog/CatalogTestBase.java:606]: Why changing this?

**Review by tsreaper — CHANGES_REQUESTED** [review]

**Comment by tsreaper** [comment]
I would also prefer an IT case for `ALTER TABLE ... MODIFY` between all possible types. You mistakenly use `String` and this should be easily discovered by IT cases.

File: paimon-core/src/test/java/org/apache/paimon/catalog/CatalogTestBase.java (hunk 5 above)
Comment (zhuangchong) [reply paimon-core/src/test/java/org/apache/paimon/catalog/CatalogTestBase.java:606]: This test case is Alter table update a column type throws Exception when column data type does not support cast, this PR is modified to support BIGINT to STRING, this case is correct, so I changed it to BIGINT to DATE.

File: paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java (hunk 1 above)
Comment (zhuangchong) [reply paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java]: done.

File: paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java (hunk 2 above)
Comment (zhuangchong) [reply paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java]: done.

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaManager.java (hunk 3 above)
Comment (zhuangchong) [reply paimon-core/src/main/java/org/apache/paimon/schema/SchemaManager.java:286]: This PR is to quickly solve the known BUG of users, support bigint conversion string, I have created a new issue to supports all explicit casting.

File: paimon-core/src/test/java/org/apache/paimon/casting/CastExecutorTest.java (hunk 4 above)
Comment (zhuangchong) [reply paimon-core/src/test/java/org/apache/paimon/casting/CastExecutorTest.java:148]: done.

**Comment by zhuangchong** [comment]
> I would also prefer an IT case for `ALTER TABLE ... MODIFY` between all possible types. You mistakenly use `String` instead of `BinaryString` and this should be easily discovered by IT cases.

@tsreaper 
Thank you very much for your code review, I have modified it and added a test case.

https://github.com/apache/incubator-paimon/blob/7430f0456f84badaac7deca8d5284b9565dd97e6/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java#L210-L226

**Review by tsreaper — COMMENTED** [review]
Never squash and force-push your commit unless you have to resolve conflicts by rebasing. If you squash and force-push I can't see what has changed since my last review.

File: paimon-core/src/test/java/org/apache/paimon/catalog/CatalogTestBase.java (hunk 5 above)
Comment (tsreaper) [reply paimon-core/src/test/java/org/apache/paimon/catalog/CatalogTestBase.java:606]: Each PR should be complete by its own. This PR is not for solving bugs, but for introducing a new feature. By introducing this new feature do you really want to create new bugs and just leave them behind? Please fix this.

File: paimon-common/src/main/java/org/apache/paimon/utils/TypeUtils.java [hunk 6]
```diff
@@ -170,6 +170,10 @@ public static boolean toBoolean(BinaryString str) {
         throw new RuntimeException("Cannot parse '" + str + "' as BOOLEAN.");
     }
 
+    public static int toInt(Boolean bool) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/utils/TypeUtils.java:173]: `Boolean` -> `boolean`?

File: paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java [hunk 7]
```diff
@@ -215,19 +245,52 @@ public class CastExecutors {
                                             (((Timestamp) value).getMillisecond()
                                                     % DateTimeUtils.MILLIS_PER_DAY);
                         }
+                    case TINYINT:
+                        return value -> (byte) (((Timestamp) value).getMillisecond());
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java:249]: Is `getMillisecond` correct? You can take a look to Flink and Spark.

File: paimon-common/src/main/java/org/apache/paimon/utils/TypeUtils.java (hunk 6 above)
Comment (zhuangchong) [reply paimon-common/src/main/java/org/apache/paimon/utils/TypeUtils.java]: done.

File: paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java (hunk 7 above)
Comment (zhuangchong) [reply paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java]: done.

File: paimon-core/src/test/java/org/apache/paimon/catalog/CatalogTestBase.java (hunk 5 above)
Comment (zhuangchong) [reply paimon-core/src/test/java/org/apache/paimon/catalog/CatalogTestBase.java:606]: New support for type cast: String to Number, Boolean to Number or String, Timestamp to Number or String.

**Comment by zhuangchong** [comment]
The exception caused the CI exception when downloading the jar, and I restarted the PR multiple times to trigger CI to rerun.

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Hi @zhuangchong , thanks for the contribution.
Left comments.
You need to add more tests, especially for round-trip testing, we need to cover all round-trip tests.
For example, timestamp cast to string, and string cast to timestamp.

The tests are very very important, please put more effort into testing, cover the tests sufficiently first, and even compare the tests in Flink.

**Review by JingsongLi — COMMENTED** [review]
Thanks @zhuangchong for updating! Very Nice!

We can document this too, in https://paimon.apache.org/docs/master/how-to/altering-tables/#changing-column-type
We can document the conversion table.

File: paimon-core/src/main/java/org/apache/paimon/casting/TimestampToStringCastRule.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                CastRulePredicate.builder()
+                        .input(DataTypeFamily.TIMESTAMP)
+                        .target(STRING_TYPE)
+                        .build());
+    }
+
+    @Override
+    public CastExecutor<Timestamp, BinaryString> create(DataType inputType, DataType targetType) {
+        final int precision = DataTypeChecks.getPrecision(inputType);
+        if (inputType.is(DataTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/casting/TimestampToStringCastRule.java:47]: What about `TIMESTAMP_WITH_LOCAL_ZONE`?

Comment (zhuangchong) [reply paimon-core/src/main/java/org/apache/paimon/casting/TimestampToStringCastRule.java]: done.

**Comment by zhuangchong** [comment]
> Thanks @zhuangchong for updating! Very Nice!
> 
> We can document this too, in https://paimon.apache.org/docs/master/how-to/altering-tables/#changing-column-type We can document the conversion table.

I have added type conversion docs.


image

File: paimon-core/src/main/java/org/apache/paimon/casting/TimestampToStringCastRule.java (hunk 8 above)
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/casting/TimestampToStringCastRule.java]: I mean we should support `TIMESTAMP_WITH_LOCAL_ZONE` to string too.

Comment (zhuangchong) [reply paimon-core/src/main/java/org/apache/paimon/casting/TimestampToStringCastRule.java]: Thanks. Already supports `TIMESTAMP_WITH_LOCAL_ZONE` type conversion.

**Comment by zhuangchong** [comment]
Please help to review this PR. Thanks @JingsongLi .

**Review by JingsongLi — APPROVED** [review]
Really thank you @zhuangchong , looks good to me now!

**Comment by zhuangchong** [comment]
@JingsongLi Thanks.

**Pull request merged**